#!/bin/bash

awk '{print}' "${11}" > "$1"
uniquely_mapped_reads=$(awk -F'\t' '/Uniquely mapped reads number/ {print $2}' "$2")
echo "The transcriptome sequencing of $3 produced $uniquely_mapped_reads uniquely aligned sequencing reads, enabling quantification of protein coding genes." >> "$1"
echo -e "\\nQuality metrics (fastQC / MultiQC) indicated:" >> "$1"
awk '{print}' "$4" >> "$1"
awk '{print}' "$5" >> "$1"
echo -e "\\nALLCatchR allocated for sample $3 the following molecular subtype:" >> "$1"
echo -e "subtype prediction\tscore\tconfidence" >> "$1"
awk -F'\t' 'NR > 1 { printf "%s\t%s\t%s\n", $3, $2, $4 }' "$6" >> "$1"
echo -e "\\nfusioncatcher / ARRIBA identified the following driver fusion candidates:" >> "$1"
echo -e "\\nFusioncatcher:" >> "$1"
echo -e "5’ gene name\t5’ chr.position\t3’ gene name\t3’chr. position\tfusion unique spanning reads"  >> "$1"
awk -F'\t' 'NR > 1 { printf "%s\t%s\t%s\t%s\t%s\n", $1, $9, $2, $10, $6 }' "$7" >> "$1"
echo -e "\\nARRIBA:" >> "$1"
echo -e "5’ gene name\t5’ chr.position\t3’ gene name\t3’chr. position\tdiscordant_mates"  >> "$1"
awk -F'\t' 'NR > 1 { printf "%s\t%s\t%s\t%s\t%s\n", $1, $5, $2, $6, $12 }' "$8" >> "$1"
echo -e "\\nRNASeqCNV identified the following karyotype:" >> "$1"
echo -e "gender\tchrom_n\talterations\tstatus\tcomments"  >> "$1"
awk -F'\t' 'NR > 1 { printf "%s\t%s\t%s\t%s\t%s\n", $2, $3, $4, $5, $6 }' "$9" >> "$1"
echo -e "Chromosome arm calls" >> "$1"
echo -e "chr\tarm\tlog2FC per arm"  >> "$1"
awk -F'\t' 'NR > 1 { printf "%s\t%s\t%s\n", $1, $2, $3 }' "${10}" >> "$1"
